x/ccv/provider/keeper/power_shaping.go

```diff
@@ -551,3 +550,121 @@ func (k Keeper) DeleteMinimumPowerInTopN(
+	}
+
+	return providerConsAddresses
+}
+
+// IsPrioritylisted returns `true` if validator with `providerAddr` has been prioritylisted on chain `consumerId`
+func (k Keeper) IsPrioritylisted(
+	ctx sdk.Context,
+	consumerId string,
+	providerAddr types.ProviderConsAddress,
+) bool {
+	store := ctx.KVStore(k.storeKey)
+	bz := store.Get(types.PrioritylistKey(consumerId, providerAddr))
+	return bz != nil
+}
+
+// DeletePrioritylist deletes all prioritylisted validators
+func (k Keeper) DeletePrioritylist(ctx sdk.Context, consumerId string) {
+	store := ctx.KVStore(k.storeKey)
+	iterator := storetypes.KVStorePrefixIterator(store, types.StringIdWithLenKey(types.PrioritylistKeyPrefix(), consumerId))
+	defer iterator.Close()
+
+	keysToDel := [][]byte{}
+	for ; iterator.Valid(); iterator.Next() {
+		keysToDel = append(keysToDel, iterator.Key())
+	}
+
+	for _, key := range keysToDel {
+		store.Delete(key)
+	}
+}
+
+// IsPrioritylistEmpty returns `true` if no validator is prioritylisted on chain `consumerId`
+func (k Keeper) IsPrioritylistEmpty(ctx sdk.Context, consumerId string) bool {
+	store := ctx.KVStore(k.storeKey)
+	iterator := storetypes.KVStorePrefixIterator(store, types.StringIdWithLenKey(types.PrioritylistKeyPrefix(), consumerId))
+	defer iterator.Close()
+
+	return !iterator.Valid()
+}
+
+// UpdatePrioritylist populates the prioritylist store for the consumer chain with this consumer id
+func (k Keeper) UpdatePrioritylist(ctx sdk.Context, consumerId string, prioritylist []string) {
+	k.DeletePrioritylist(ctx, consumerId)
+	for _, address := range prioritylist {
+		consAddr, err := sdk.ConsAddressFromBech32(address)
+		if err != nil {
+			continue
+		}
+
+		k.SetPrioritylist(ctx, consumerId, types.NewProviderConsAddress(consAddr))
+	}
+}
+
+// FilterAndSortPriorityList filters the priority list to include only validators that can validate the chain
+// and splits the validators into priority and non-priority sets.
+func (k Keeper) FilterAndSortPriorityList(ctx sdk.Context, consumerId string, nextValidators []types.ConsensusValidator) ([]types.ConsensusValidator, []types.ConsensusValidator) {
+	validatorMap := make(map[string]types.ConsensusValidator)
+	for _, v := range nextValidators {
+		validatorMap[string(v.ProviderConsAddr)] = v
+	}
+
+	priorityValidators := make([]types.ConsensusValidator, 0)
+	nonPriorityValidators := make([]types.ConsensusValidator, 0)
+	addedAddresses := make(map[string]bool)
+
+	// Form priorityValidators
+	for _, validator := range nextValidators {
+		addr := providertypes.NewProviderConsAddress(validator.ProviderConsAddr)
+		if k.IsPrioritylisted(ctx, consumerId, addr) {
+			if !addedAddresses[string(validator.ProviderConsAddr)] {
+				priorityValidators = append(priorityValidators, validator)
+				addedAddresses[string(validator.ProviderConsAddr)] = true
+			}
+		}
+	}
+
+	// Add remaining validators to nonPriorityValidators
+	for _, validator := range nextValidators {
```

Comment:
nit: Couldn't this be done above inside the previous `for` loop, to have an `else // !k.isPriorityListed`?

Reply:
Yes, I removed it and created the non-priority list in the same `for` as the priority list.